indexbs/indexbacked_bs.go

```diff
@@ -169,59 +169,57 @@ func (ro *IndexBackedBlockstore) execOp(ctx context.Context, c cid.Cid, op Block
 
 	// Some retrieval patterns will result in multiple threads fetching blocks
 	// from the same piece concurrently. In that case many threads may attempt
-	// to create a blockstore over the same piece. Use a sync.Once to ensure
+	// to create a blockstore over the same piece. Use a striped lock to ensure
 	// that the blockstore is only created once for all threads waiting on the
 	// same shard.
-	bsAcquireI, _ := ro.bsAcquireByShard.LoadOrStore(sk, &blockstoreAcquire{})
-	bsAcquire := bsAcquireI.(*blockstoreAcquire)
-	bsAcquire.once.Do(func() {
-		bsAcquire.bs, bsAcquire.err = func() (dagstore.ReadBlockstore, error) {
-			// Check if the blockstore was created by another thread while this
-			// thread was waiting to enter the sync.Once
-			val, ok := ro.blockstoreCache.Get(sk)
-			if ok {
-				return val.(*accessorWithBlockstore).bs, nil
-			}
-
-			// Acquire the blockstore for the selected shard
-			resch := make(chan dagstore.ShardResult, 1)
-			if err := ro.d.AcquireShard(ro.ctx, sk, resch, dagstore.AcquireOpts{}); err != nil {
-				return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, err)
-			}
-			var shres dagstore.ShardResult
-			select {
-			case <-ctx.Done():
-				return nil, ctx.Err()
-			case shres = <-resch:
-				if shres.Error != nil {
-					return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, shres.Error)
-				}
-			}
+	bs, err := func() (dagstore.ReadBlockstore, error) {
+		// Derive the striped lock index from the shard key and acquire the lock
+		skstr := sk.String()
+		lockIdx := skstr[len(skstr)-1]
```

Comment:
I am not sure why we are basing the lock index on the length of the `sk` string?